### PR TITLE
feat: BL-15039 add static analysis and linting

### PR DESCRIPTION
**Pull request by waringr**
## Description

Enables actions for static analysis and linting in CI workflow 

Upgrade to PHP 8.2

Related issue: https://dvsa.atlassian.net/browse/BL-15039

## Before submitting (or marking as "ready for review")

- [x] Does the pull request title follow the [conventional commit](https://www.conventionalcommits.org/en/v1.0.0/) specification?
- [ ] Have you performed a self-review of the code?
- [ ] Have you have added tests that prove the fix or feature is effective and working?
- [ ] Did you make sure to update any documentation relating to this change?

### CODE REVIEW

File: composer.json
Comment (UndermountainK) [inline composer.json]: Can we match the 8.2 php version?
```suggestion
        "dvsa/mot-cpms-notifications": "dev-feature-bl-17032-php-82-update",
```

File: composer.json
Comment (UndermountainK) [inline composer.json]: any reason why was it moved from require-dev?

File: phpstan.neon
Comment (UndermountainK) [inline phpstan.neon:3]: Can we add comments why the suppresions are needed?

File: src/CpmsClient/Data/AccessToken.php
Comment (CKindred) [inline src/CpmsClient/Data/AccessToken.php:30]: How come this ignore was needed?

File: src/CpmsClient/Client/ClientOptions.php
Comment (UndermountainK) [inline src/CpmsClient/Client/ClientOptions.php]: Minor: We can swap array() for []
```suggestion
    protected array $endPoints = [];
```

File: src/CpmsClient/Client/RestClientFactory.php
Comment (UndermountainK) [inline src/CpmsClient/Client/RestClientFactory.php]: Would it worked if we type @param $requestedName as mixed?

File: src/CpmsClient/Service/ApiService.php
Comment (UndermountainK) [inline src/CpmsClient/Service/ApiService.php:58]: Will this construct change the behaviour? if not can we use here "Constructor property promotion"?

File: src/CpmsClient/Service/ApiService.php
Comment (UndermountainK) [inline src/CpmsClient/Service/ApiService.php:113]: Was that code not needed?

File: src/CpmsClient/Service/CacheAwareApiService.php
Comment (UndermountainK) [inline src/CpmsClient/Service/CacheAwareApiService.php]: Can we use here "Constructor property promotion"?

File: test/ApplicationTest/Service/CacheAwareApiServiceTest.php
Comment (UndermountainK) [inline test/ApplicationTest/Service/CacheAwareApiServiceTest.php:1]: Moved file to different dir was not needed. Better to change namespace if that was purpose of this.

File: test/Bootstrap.php
Comment (UndermountainK) [inline test/Bootstrap.php:4]: Does "init_bootstrap.php" exists? Was it intended?
EDIT: Okay it was added. Ignore this comment :)

File: test/CpmsClientTest/Bootstrap.php
Comment (UndermountainK) [inline test/CpmsClientTest/Bootstrap.php:102]: Does something like:
```
array_key_exists('listeners', $config) &&
            is_array($config['listeners'])
```
would make suppression not needed?

File: test/CpmsClientTest/Bootstrap.php
Comment (UndermountainK) [inline test/CpmsClientTest/Bootstrap.php:150]: Was it intended? it seems like this was needed
EDIT: Okay it was moved to the "test/CpmsClientTest/init_bootstrap.php". Ignore this comment :)

File: test/CpmsClientTest/MockApiService.php
Comment (UndermountainK) [inline test/CpmsClientTest/MockApiService.php]: i am not a fan of using "#[AllowDynamicProperties]". Usually it comes up to declaring class variables for those using class scope like: private $var, when in code we declare $this->var...

**Review by UndermountainK — CHANGES_REQUESTED** [review]
Great job Alexis. Left come comments.
Wehn dvsa/mot-cpms-notifications will be changed to match 8.2 i can test it in swimlane if its working.

**Review by CKindred — DISMISSED** [review]
LGTM pending @UndermountainK's suggested changes

File: src/CpmsClient/Service/ApiService.php
Comment (alexisc-kainos) [reply src/CpmsClient/Service/ApiService.php:113]: not sure, but there seems to be no variable $data, plus data['user_id'] gives a warning "write access unused"

File: composer.json
Comment (alexisc-kainos) [reply composer.json]: changed back

File: src/CpmsClient/Client/ClientOptions.php
Comment (alexisc-kainos) [reply src/CpmsClient/Client/ClientOptions.php]: done

File: src/CpmsClient/Client/RestClientFactory.php
Comment (alexisc-kainos) [reply src/CpmsClient/Client/RestClientFactory.php]: it does :)

File: src/CpmsClient/Service/ApiService.php
Comment (alexisc-kainos) [reply src/CpmsClient/Service/ApiService.php:58]: Have changed to constructor property promotion

File: src/CpmsClient/Service/CacheAwareApiService.php
Comment (alexisc-kainos) [reply src/CpmsClient/Service/CacheAwareApiService.php]: changed

File: test/ApplicationTest/Service/CacheAwareApiServiceTest.php
Comment (alexisc-kainos) [reply test/ApplicationTest/Service/CacheAwareApiServiceTest.php:1]: woops, think that was an intelliJ auto-click, moved back

File: test/CpmsClientTest/Bootstrap.php
Comment (alexisc-kainos) [reply test/CpmsClientTest/Bootstrap.php:102]: Not sure what you mean?

File: test/CpmsClientTest/MockApiService.php
Comment (alexisc-kainos) [reply test/CpmsClientTest/MockApiService.php]: removed

File: phpstan.neon
Comment (alexisc-kainos) [reply phpstan.neon:3]: yeah, that's a good point, I think the reason might just be due to the fact that this rule was introduced after the code was written so there are a lot of instances of "no value specified on iterable type array". The suppression was already there when we added the linting (and the same in all the other tickets). 

I'm working on fixing the issues that are suppressed

Comment (alexisc-kainos) [reply phpstan.neon:3]: We've decided that to fix the issues highlighted by this suppression is a wide-reaching change - adding typing would require a new ticket to investigate in more detail. We will leave as-is for now 
ticket created: https://dvsa.atlassian.net/browse/BL-17327

File: test/CpmsClientTest/Bootstrap.php
Comment (UndermountainK) [reply test/CpmsClientTest/Bootstrap.php:102]: Checked and it does not fix the need for suppression. Resolved

File: src/CpmsClient/Service/ApiService.php
Comment (alexisc-kainos) [reply src/CpmsClient/Service/ApiService.php:113]: re-added code, it needed a different parameter name $params.

File: src/CpmsClient/Client/ClientOptions.php
Comment (waringr) [inline src/CpmsClient/Client/ClientOptions.php]: Does it?

File: src/CpmsClient/Data/AccessToken.php
Comment (waringr) [inline src/CpmsClient/Data/AccessToken.php:28]: Can we limit this to just the rule that's failing? So we don't accidentally mask something else

Comment (alexisc-kainos) [reply src/CpmsClient/Data/AccessToken.php:28]: there's no specific rule, I'd have to add something to the php neon like: 
ignoreErrors:
        - message: '#Parameter #1 \$options of method Laminas\\Stdlib\\AbstractOptions<iterable>::__construct\(\) expects iterable<string, iterable>\|Laminas\\Stdlib\\AbstractOptions<iterable>\|null, iterable\|null given.#'

I thought this way looked better

File: src/CpmsClient/Client/ClientOptions.php
Comment (alexisc-kainos) [reply src/CpmsClient/Client/ClientOptions.php]: it does not, fixed

**Comment by waringr** [comment]
I can't approve, but LGTM

**Review by UndermountainK — APPROVED** [review]

**Review by CKindred — APPROVED** [review]

**Review by aaronlong — APPROVED** [review]

**Pull request merged**
